newshomepages/save_html.py

```diff
@@ -0,0 +1,130 @@
+import typing
+from pathlib import Path
+
+import click
+from playwright.sync_api import sync_playwright
+from retry import retry
+from rich import print
+
+from . import utils
+
+
+@click.command()
+@click.argument("handle")
+@click.option("-o", "--output-dir", "output_dir", default="./")
+@click.option("-w", "--wait", "wait", default=5000)
+@click.option("-x", "--width", "width", default=1300)
+@click.option("-y", "--height", "height", default=1600)
+@click.option(
+    "-f",
+    "--full-page",
+    "full_page",
+    is_flag=True,
+    default=False,
+    help="Lazy load all images, just to make sure they're in the final single-page html",
+)
+def cli(
+    handle: str,
+    output_dir: str = "./",
+    wait: str = "5000",
+    width: str = "1300",
+    height: str = "1600",
+    full_page: bool = False,
+):
+    """Save HTML for the provided homepage."""
+    site = utils.get_site(handle)
+    output_path = Path(output_dir)
+    _save_html(
+        site,
+        output_path,
+        wait=int(wait),
+        width=int(width),
+        height=int(height),
+        full_page=bool(full_page),
+    )
+
+
+@retry(tries=3, delay=5, backoff=2)
+def _save_html(
+    site: typing.Dict,
+    output_path: Path,
+    wait: int = 5000,
+    width: int = 1300,
+    height: int = 1600,
+    full_page: bool = False,
+):
+    """Shoot the provided site."""
+    print(f":camera: Fetching HML for {site['name']}")
+
+    # Set the output path
+    output_path.mkdir(parents=True, exist_ok=True)
+
+    with sync_playwright() as playwright:
+        context = utils._load_persistent_context(playwright, width, height)
+        page = utils._load_new_page_disable_javascript(
+            context=context,
+            url=site["url"],
+            wait_seconds=int((site["wait"] or wait) / 1000),
+            handle=site["handle"],
+            full_page=full_page,
+        )
```

Comment:
Do we need to run the JavaScript filters?

Reply:
Yes, pop-ups and paywalls are saved to the single-page HTML. See here (screenshotted)

<img width="1725" alt="temp-latimes-screenshot" src="https://user-images.githubusercontent.com/3460632/197872915-a53bd1d8-894e-492c-b2aa-5b2e6a9db959.png">

Reply:
not ideal.. and that's WITH the `disable-javascript` running. Somehow it seems to block the LATimes images from being saved. NYTimes and WP don't have those issues AFAICT:

<img width="1423" alt="temp-nyt-html" src="https://user-images.githubusercontent.com/3460632/197873344-d7c8f021-2873-432d-a47d-d3f0fbe850cd.png">

<img width="1703" alt="temp-wp-screenshot" src="https://user-images.githubusercontent.com/3460632/197873345-51b97a02-b265-4617-ace5-86c98b0c52b3.png">

Reply:
I'm not sure why HTML saves well for some websites but not for others. 

But we don't actually need a huge set of these websites to be saving full-page HTML. 

All we need is a few that we craft custom CSS for and use as training.